examples/Nat/NAT.cs

```diff
@@ -0,0 +1,144 @@
+using System.Timers;
+using PacketDotNet;
+
+namespace Pax.Examples.Nat
+{
+  /// <summary>
+  /// A packet processor that performs Network Address Translation (a NAT).
+  /// </summary>
+  public sealed class NAT : SimplePacketProcessor
+  {
+    /// A value indicating the packet should be dropped.
+    public const int Port_Drop = -1;
+
+    private readonly Timer gcTimer;
+
+    // Use a separate namespace for each transport protocol
+    private TcpNAT tcpNat;
+    private UdpNAT udpNat;
+
+    /// <summary>
+    /// Creates a new NAT packet processor that handles TCP and UDP packets.
+    /// </summary>
+    /// <param name="my_address">The public IP address of the NAT</param>
+    /// <param name="next_outside_hop_mac">The MAC address of the next hop on the outside-facing port.</param>
+    /// <param name="tcp_inactivity_timeout">The time that should elapse before a TCP connection with no activity is removed.</param>
+    /// <param name="tcp_time_wait_duration">The time that should elapse before a closed TCP connection is removed.</param>
+    /// <param name="tcp_start_port">The start of the range of TCP ports to use (inclusive).</param>
+    /// <param name="tcp_end_port">The end of the range of TCP ports to use (inclusive).</param>
+    /// <param name="udp_inactivity_timeout">The time that should elapse before a UDP connection with no activity is removed.</param>
+    /// <param name="udp_start_port">The start of the range of UDP ports to use (inclusive).</param>
+    /// <param name="udp_end_port">The end of the range of UDP ports to use (inclusive).</param>
+    public NAT (IPAddress my_address, PhysicalAddress next_outside_hop_mac,
+      TimeSpan tcp_inactivity_timeout, TimeSpan tcp_time_wait_duration, ushort tcp_start_port, ushort tcp_end_port,
+      TimeSpan udp_inactivity_timeout, ushort udp_start_port, ushort udp_end_port)
+    {
+      // Instantiate each NAT specialisation
+      tcpNat = new TcpNAT(my_address, next_outside_hop_mac, tcp_inactivity_timeout, tcp_time_wait_duration, tcp_start_port, tcp_end_port);
+      udpNat = new UdpNAT(my_address, next_outside_hop_mac, udp_inactivity_timeout, udp_start_port, udp_end_port);
+
+      // Call the GarbageCollectConnections method regularly
+      gcTimer = new Timer(1000); // FIXME should the GC frequency be configurable?
+      gcTimer.Elapsed += GarbageCollectConnections;
+      gcTimer.AutoReset = true;
+      gcTimer.Start();
+    }
+
+    /// <summary>
+    /// Handle an observed packet. Determines the type of packet and passes it to the relevant specialised handler.
+    /// </summary>
+    /// <param name="incomingNetworkInterface">The number representing network interface the packet arrived on.</param>
+    /// <param name="packet">The observed packet.</param>
+    /// <returns>A <see cref="ForwardingDecision"/> for the packet.</returns>
+    override public ForwardingDecision process_packet(int incomingNetworkInterface, ref Packet packet)
```

Comment:
Nice, I like how this function makes it easy for some core logic to bubble up to the surface of the code, and protocol-specific handling (e.g., for TCP) is done in dedicated handlers.